### PR TITLE
fix: touch event canceled with gesture disabled

### LINKED ISSUE #1022: RN Touch events are canceled even when {gestureEnabled: false} on iOS when using native stack navigation
## Description

For example, if you have a component at the edge like a slider, it's touches are canceled if you start sliding from the left edge. That might be alright if the screen has gesture enabled, but I don't think it's correct to do that when gestureEnabled: false for a screen (totally possible I am missing a case).

I haven't worked with Objective C before but I think the issue is because of this.

https://github.com/software-mansion/react-native-screens/blob/c3998ad5f95d2bbd881d19bce4de52b151fafde4/ios/RNSScreenStack.m#L530-L553

Specifically, the part where touchHandler is cancelled and reset without checking if gesture's enabled for a screen.

I fixed it by applying this patch locally that tests if topScreen.gestureEnabled is truthy before cancelling/resetting touch handlers :
```
@@
diff --git a/node_modules/react-native-screens/ios/RNSScreenStack.m b/node_modules/react-native-screens/ios/RNSScreenStack.m
index 7b76e12..f44f11a 100644
--- a/node_modules/react-native-screens/ios/RNSScreenStack.m
+++ b/node_modules/react-native-screens/ios/RNSScreenStack.m
@@ -509,20 +509,22 @@

 - (BOOL)gestureRecognizerShouldBegin:(UIGestureRecognizer *)gestureRecognizer
 {
-  // cancel touches in parent, this is needed to cancel RN touch events. For example when Touchable
-  // item is close to an edge and we start pulling from edge we want the Touchable to be cancelled.
-  // Without the below code the Touchable will remain active (highlighted) for the duration of back
-  // gesture and onPress may fire when we release the finger.
-  UIView *parent = _controller.view;
-  while (parent != nil && ![parent respondsToSelector:@selector(touchHandler)]) parent = parent.superview;
-  if (parent != nil) {
-    RCTTouchHandler *touchHandler = [parent performSelector:@selector(touchHandler)];
-    [touchHandler cancel];
-    [touchHandler reset];
-  }
-  
   RNSScreenView *topScreen = (RNSScreenView *)_controller.viewControllers.lastObject.view;

+  if(topScreen.gestureEnabled) {
+    // cancel touches in parent, this is needed to cancel RN touch events. For example when Touchable
+    // item is close to an edge and we start pulling from edge we want the Touchable to be cancelled.
+    // Without the below code the Touchable will remain active (highlighted) for the duration of back
+    // gesture and onPress may fire when we release the finger.
+    UIView *parent = _controller.view;
+    while (parent != nil && ![parent respondsToSelector:@selector(touchHandler)]) parent = parent.superview;
+    if (parent != nil) {
+      RCTTouchHandler *touchHandler = [parent performSelector:@selector(touchHandler)];
+      [touchHandler cancel];
+      [touchHandler reset];
+    }
+  }
+  
   if (!topScreen.gestureEnabled || _controller.viewControllers.count < 2) {
     return NO;
   }
   ```
## Steps To Reproduce
1. Use native stack with gestureEnabled: false on iOS
2. Use a slider. For ex [this](https://www.npmjs.com/package/@sharcoux/slider), just import and use `<Slider maximumValue={100} />`
3. Start sliding from the left edge, you'll see that the slider is interrupted

### Expected behavior
Touch handler shouldn't be interrupted when gestureEnabled:false

### Actual behavior
Touch handler is interrupted
## Package versions

- React Native: 0.63.4
- React Native Screens: 3.4.0

### PR DESCRIPTION
**Pull request by Ubax**
## Description
Fixes #1022

## Changes
Adds event cancelation only when gesture should be recognized.

## Test code and steps to reproduce
Test1032.tsx

## Checklist

- [x] Included code example that can be used to test this change
- [x] Ensured that CI passes

### CODE REVIEW

File: TestsExample/src/Test1032.tsx [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+const Stack = createNativeStackNavigator();
+
+export default function App() {
+    const [gestureEnabled, setGestureEnable] = React.useState(false);
+    return (
+        <View style={{ flex: 1, paddingBottom: 200 }}>
+            <NavigationContainer>
+                <Stack.Navigator screenOptions={{ gestureEnabled }}>
+                    <Stack.Screen name="Top" component={First} options={{ replaceAnimation: 'push' }} />
```
Comment (WoLewicki) [inline TestsExample/src/Test1032.tsx:14]: `options={{ replaceAnimation: 'push' }}` is this option needed for something?

File: TestsExample/src/Test1032.tsx [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+            <Button title="Tap me for second screen" onPress={() => navigation.push('Top1')} />
+        </View>
+
+    );
+}
+
+function Second() {
+    return (
+        <View style={{ backgroundColor: '#FFF' }}>
+            <Button title="Tap me to see test" onPress={() => Alert.alert('Test')} />
```
Comment (WoLewicki) [inline TestsExample/src/Test1032.tsx:42]: Maybe better something like `Swipe back on me to see if button triggers`

File: TestsExample/src/Test1032.tsx [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+        <View style={{ flex: 1, paddingBottom: 200 }}>
+            <NavigationContainer>
+                <Stack.Navigator screenOptions={{ gestureEnabled }}>
+                    <Stack.Screen name="Top" component={First} options={{ replaceAnimation: 'push' }} />
+                    <Stack.Screen name="Top1" component={Second} options={{ replaceAnimation: 'push' }} />
+                </Stack.Navigator>
+            </NavigationContainer >
+            <View>
+                <Text>Gesture enabled</Text>
+                <Switch value={gestureEnabled} onValueChange={setGestureEnable} />
```
Comment (WoLewicki) [inline TestsExample/src/Test1032.tsx:20]: You can switch `gestureEnabled` directly in `Second` by adding such switch there and using `navigation.setOptions`.

File: ios/RNSScreenStack.m [hunk 4]
```diff
@@ -526,7 +526,7 @@ - (void)dismissOnReload
   return nil;
 }
 
-- (BOOL)gestureRecognizerShouldBegin:(UIGestureRecognizer *)gestureRecognizer
+- (void)cancelTouchHandlerIfExists
```
Comment (WoLewicki) [inline ios/RNSScreenStack.m:529]: I think `cancelTouchesInParent` fits better according to the comment.

File: ios/RNSScreenStack.m [hunk 5]
```diff
@@ -556,13 +560,16 @@ - (BOOL)gestureRecognizerShouldBegin:(UIGestureRecognizer *)gestureRecognizer
                           ((RNSGestureRecognizer *)gestureRecognizer).edges == UIRectEdgeRight) ||
         (_controller.view.semanticContentAttribute != UISemanticContentAttributeForceRightToLeft &&
          ((RNSGestureRecognizer *)gestureRecognizer).edges == UIRectEdgeLeft);
-    if (isCorrectEdge) {
-      return topScreen.stackAnimation == RNSScreenStackAnimationSimplePush;
+    if (isCorrectEdge && topScreen.stackAnimation == RNSScreenStackAnimationSimplePush) {
+      [self cancelTouchHandlerIfExists];
+      return YES;
     }
     return NO;
```
Comment (WoLewicki) [inline ios/RNSScreenStack.m:567]: this `return NO;` seems unnecessary here.

**Review by WoLewicki — APPROVED** [review]
Looks good! I left some comments, please answer them and I think we can merge it then.

File: TestsExample/src/Test1032.tsx (hunk 1 above)
Comment (Ubax) [reply TestsExample/src/Test1032.tsx:14]: No. Actually it is a leftover

File: TestsExample/src/Test1032.tsx (hunk 3 above)
Comment (Ubax) [reply TestsExample/src/Test1032.tsx:20]: I think that being able to change it on both screens is useful, so I would keep it outside

**Review by kacperkapusciak — APPROVED** [review]
LGTM 👏

**Pull request merged**